charts/openstack-cluster/DEBUGGING.md

```diff
@@ -0,0 +1,172 @@
+```
+
+The progress of these steps can also be monitored using the OpenStack Horizon dashboard or the `openstack` CLI.
+
+Commonly encountered issues at this stage (usually visible in the CAPO controller logs) include:
+
+- Incorrectly formatted cloud credentials (see [these docs](./README.md#openstack-credentials))
+
+- Insufficient quota in the target OpenStack project (for various resources such as networks, subnets, security groups etc.)
+
+- Incorrect network names, IDs or other filters (e.g. if both an ID and a name are provided simultaneously in network filter config then CAPO might do [unexpected things](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/6560f8882a2aa7ece3d13d47f2f2badbcba348c3/api/v1beta1/types.go#L160))
+
+### First control plane node
+
+After creating any necessary networking resources, the CAPO controller will create the first control plane node. Once the node has initialised successfully, the output of the `kubeadm init` command should be visible in the server logs (visible e.g. in the OpenStack Horizon dashboard):
+
+```
+ Your Kubernetes control-plane has initialized successfully!
+
+ To start using your cluster, you need to run the following as a regular user:
+
+   mkdir -p $HOME/.kube
+   sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
+   sudo chown $(id -u):$(id -g) $HOME/.kube/config
+
+ Alternatively, if you are the root user, you can run:
+
+   export KUBECONFIG=/etc/kubernetes/admin.conf
+
+ You should now deploy a pod network to the cluster.
+.yaml" with one of the options listed at:
+   https://kubernetes.io/docs/concepts/cluster-administration/addons/
+
+ You can now join any number of control-plane nodes by copying certificate authorities
+ and service account keys on each node and then running the following as root:
+
+   kubeadm join <ip-address>:6443 --token <join-token> \
+ 	--discovery-token-ca-cert-hash <hash> \
+ 	--control-plane
+```
+
+> [!NOTE]
+> No manual kubeadm commands are required as the remaining cluster nodes will join the cluster automatically. The above output is just the standard help text output from `kubeadm`.
+
+Commonly encountered issues (usually visible in the CAPO controller logs) include:
+
+- Insufficient quota in the target OpenStack project (for servers, volumes etc.)
+
+- Network routing issues between the CAPI management cluster and the workload cluster
+
+### Essential Addons
+
+Now that the first control plane node has initialised successfully, the kubeconfig for the workload cluster will be written to a Kubernetes `Secret` named `<cluster-name>-kubeconfig` on the CAPI management cluster. At this point, the kubeconfig can be extracted from the secret for use with `kubectl` using the following command:
+
+```
+kubectl --kubeconfig capi-mgmt get secret <cluster-name>-kubeconfig -o jsonpath='{.data.value}' | base64 -d > workload
```

Comment:
The `go-template` format for `kubectl` supports a `base64decode` filter which I think is nicer than this:

```sh
kubectl --kubeconfig capi-mgmt get secret <cluster-name>-kubeconfig -o go-template='{{.data.value | base64decode}}' > workload
```

Not massively bothered either way though.